src/c-writer.cc

```diff
@@ -678,50 +721,50 @@ void CWriter::Write(const LocalName& name) {
   Write(local_sym_map_[name.name]);
 }
 
-std::string CWriter::GetGlobalName(const std::string& name) const {
-  assert(global_sym_map_.count(name) == 1);
-  auto iter = global_sym_map_.find(name);
+template <ModuleFieldType T>
+void CWriter::Write(const GlobalName<T>& name) {
+  std::string mangled = name.name + MangleField(T);
+  assert(global_sym_map_.count(mangled) == 1);
+  auto iter = global_sym_map_.find(mangled);
   assert(iter != global_sym_map_.end());
-  return iter->second;
-}
-
-void CWriter::Write(const GlobalName& name) {
-  Write(GetGlobalName(name.name));
+  Write(iter->second);
 }
 
-void CWriter::Write(const ExternalPtr& name) {
+template <ModuleFieldType T>
+void CWriter::Write(const ExternalPtr<T>& name) {
   bool is_import = import_syms_.count(name.name) != 0;
-  if (is_import) {
-    Write(GetGlobalName(name.name));
-  } else {
-    Write(AddressOf(GetGlobalName(name.name)));
+  if (!is_import) {
+    Write("&");
   }
+  Write(GlobalName<T>(name.name));
 }
 
-void CWriter::Write(const ExternalInstancePtr& name) {
+template <ModuleFieldType T>
+void CWriter::Write(const ExternalInstancePtr<T>& name) {
   bool is_import = import_syms_.count(name.name) != 0;
-  if (is_import) {
-    Write("instance->", GetGlobalName(name.name));
-  } else {
-    Write("&instance->", GetGlobalName(name.name));
+  if (!is_import) {
+    Write("&");
   }
```

Comment:
How about splitting out the removal of `Deref` and `AddressOf` too.. that seems independent maybe?

Reply:
Split out to #2041.